### ISSUE TITLE
GC Can delete that are not supposed to be deleted. 

### ISSUE THREAD

**Issue by Vad1mo** [goharbor/harbor#20547]
**Expected behavior and actual behavior:**
We have observed that a few times already, In some cases the GC might delete images that were not scheduled to be deleted. 
The result is that information is present in Harbor UP and DB but not in S3. 

It is also observable in the GC Logs that the manifest was deleted. 

**Steps to reproduce the problem:**

In the UI the image is still visible
![image](https://github.com/goharbor/harbor/assets/1492007/85cb09e9-876d-47a2-a25a-17496474b112)

The image manifest sha is start with **d4f9a6cf**

```shell
#GC LOG
2024-06-04T04:13:09Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:238]: blob eligible for deletion: sha256:d4f9a6cf78a2482148fd3a429c1d2019bf27a3cee1dc74856344a5e03c521585

2024-06-04T04:14:30Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:366]: [108/1438] delete blob from storage: sha256:d4f9a6cf78a2482148fd3a429c1d2019bf27a3cee1dc74856344a5e03c521585
2024-06-04T04:14:30Z [INFO] [/jobservice/job/impl/gc/garbage_collection.go:395]: [108/1438] delete blob record from database: 5040, sha256:d4f9a6cf78a2482148fd3a429c1d2019bf27a3cee1dc74856344a5e03c521585

```
trying to pull this image results in `manifest unknown` instead of `not found` if the images don't exist anymore.
```shell


We observed that during the GC run there have been some other operations going on DB level, indicating the we have run out of DB connections. 

```sh
2024-06-04T04:22:39Z [ERROR] [/pkg/notifier/notifier.go:203]: Error occurred when triggering handler *artifact.Handler of topic PUSH_ARTIFACT: failed to connect to `host=harbor-pg-database user=harbor database=harbor`: server error (FATAL: remaining connection slots are reserved for non-replication superuser connections (SQLSTATE 53300))
``` 

**Versions:**
- harbor version: 2.7.x, 2.9.x, 2.10.

**Additional context:**

- **Harbor config files:** You can get them by packaging `harbor.yml` and files in the same directory, including subdirectory.
- **Log files:** You can get them by package the `/var/log/harbor/` .

**Comment by Vad1mo** [comment]
maybe related to https://github.com/goharbor/harbor/pull/20452

**Comment by wy65701436** [comment]
Similar issue: https://github.com/goharbor/harbor/issues/19401

**Comment by wy65701436** [comment]
The issue may be caused by the beego ORM, as it doesn't carry errors during data scanning. In some cases, such as when a connection is out of use, the ORM returns incorrect data, leading to wrong blob deletion candidates. We're working on upgrading Beego with this pull request. https://github.com/goharbor/harbor/pull/20555

To mitigate the issue, you can now schedule garbage collection during low usage time slots.

**Comment by zyyw** [comment]
> maybe related to [beego/beego#5255](https://github.com/beego/beego/issues/5255) resolved by #20452

We have this PR for it:
* https://github.com/goharbor/harbor/pull/20555

**Comment by hudymi** [comment]
@Vad1mo, does the Harbor 2.12 with the fix solve this issue for you? We are experiencing the same issue, and we have lots of containers broken because of missing layers.

**Comment by wy65701436** [comment]
@hudymi yes, it's in the v2.12. Can you confirm whether the missing layers are removed by GC? If so, could you also confirm that the removed digests do not belong to any artifacts in use at the time of GC execution?

**Comment by hudymi** [comment]
@wy65701436 are there any steps that I should follow to check the compatibility? Our check was simply based on docker pull command and then checking if missing layer is in S3 (where it was missing)

**Comment by hudymi** [comment]
@prgss yes

**Comment by hudymi** [comment]
One thing. We are on Harbor 2.11, and I asked if 2.12 fixes the problem, so can we reenable GC after the upgrade.

**Comment by piyush94** [comment]
@wy65701436 We upgraded from 2.10.3 to 2.12.2 and GC deleted images that had valid tags.

**Comment by dgaponcic** [comment]
Hey, we are running on harbor 2.12.2, and started seeing MANIFEST_BLOB_UNKNOWN errors. All the sha in those errors can be linked to logs in the GC about the blob being marked for deletion. Anyone managed to fix the issue?

**Comment by Vad1mo** [comment]
@dgaponcic @piyush94 need more info and recreation scenarios to track down the error
